File: Stack/224. Basic Calculator/Solution.py
class Solution(object):
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        n = len(s)
        stack = []
        result, number, sign=0,0,1
        for c in s:
            if c.isdigit():
                number = number*10+int(c)
            elif c=='(':
                stack.append(result)
                stack.append(sign)
                result, number, sign=0,0,1
            elif c==')':
                result += sign*number
                result *= stack.pop()
                result += stack.pop()
                number, sign = 0, 1
            elif c=='+':
                result += sign*number
                number, sign = 0, 1
            elif c=='-':
                result += sign*number
                number, sign = 0, -1
        if number>0:
            result += sign*number
        return result
    

File: Stack/224. Basic Calculator/Solution.java
class Solution {
    //(,),+,-,space,non-negative number
    public int calculate(String s) {
        int n = s.length();
        Deque<Integer> stack = new ArrayDeque<>();
        int result=0, sign=1, number=0;
        for (int i=0; i<n; i++){
            char c = s.charAt(i);
            if (Character.isDigit(c)){
                number = number*10 + (c-'0');
            }else if (c=='('){
                stack.offerFirst(result);
                stack.offerFirst(sign);
                number=0;
                result=0;
                sign=1;
            }else if (c==')'){
                result += sign*number;
                sign = stack.pollFirst();
                result = stack.pollFirst() + sign*result;
                number = 0;
                sign=1;
            }else if (c=='+'){
                result += sign*number;
                sign=1;
                number=0;
            }else if (c=='-'){
                result += sign*number;
                sign=-1;
                number=0;
            }
        }
        if (number>0){
            result += sign*number;
        }
        return result;
    }
}